src/MongoDB.Driver.Core/Core/Bindings/ReadPreferenceBinding.cs

```diff
@@ -68,15 +69,24 @@ public ICoreSessionHandle Session
         public IChannelSourceHandle GetReadChannelSource(CancellationToken cancellationToken)
         {
             ThrowIfDisposed();
-            var server = _cluster.SelectServer(_serverSelector, cancellationToken);
+            var server = ServerSelectionHelper.SelectServerAndPinIfNeeded(
+                _cluster,
+                _session,
+                _serverSelector,
+                cancellationToken);
```

Comment:
As I review this it occurs to me that pinning might more appropriately handled in a new overload of Cluster.SelectServer.

See my prototype of such a change here:

https://github.com/rstam/mongo-csharp-driver/commit/f7366ce6431a6f8f2f39fc0053b32c4eb5559a56

It results in a much less intrusive change to the bindings.